### ISSUE TITLE
No propagation of type

### ISSUE THREAD

**Issue by Kharos102** [Vector35/binaryninja-api#4582]
**Version and Platform (required):**
- Binary Ninja Version: 3.5.4480-dev, 60681746
- OS: windows
- OS Version: 11
- CPU Architecture: x86_64

**Bug Description:**
When setting a variable (var1) or argument to a different or custom type and another variable (var2) references it (e.g. `var1 = var2`, or `var1 = *var2`), the type information we set on var1 does not propagate to var2.

**Steps To Reproduce:**
Open anything (e.g. Kernelbase.dll), find a function (e.g. `PsspValidateWin32WalkMarker`) and set arg1 to a double pointer (e.g. `struct OBJECT_ATTRIBUTES** arg1`)

Now view anything that does something like 
`int32_t rbx_1 = *arg1`

Note that in this case above, rbx_1 maintained its old/original type, making this assignment a cast, when I'd want it to propagate the type I specifically set for arg1.

I don't see an option for this either? Which makes finding cross-refs for certain types difficult, as you not only have to change the type in an argument, but all variables that copy its value.

**Expected Behavior:**
I should expect to see something like:
`OBJECT_ATTRIBUTES* rbx_1 = *arg1`

**Screenshots:**
![image](https://github.com/Vector35/binaryninja-api/assets/16039802/0b695603-b92f-4810-a040-e92a03f6280d)

**Additional Information:**

**Comment by Kharos102** [comment]
Just adding this is may not be the best example I chose, but it repros in cases where the new var (e.g. `rbx_1`) is actually supposed to be based on the newly defined type (e.g. `OBJECT_ATTRIBUTES*`).

**Comment by plafosse** [comment]
Could you provide the binary? Its rather difficult to figure out whats happening here. There are a few things that can cause this. First is the confidence of the types in question. If the incoming type is lower confidence than the existing type then no types will be propagated.

**Comment by Kharos102** [comment]
Sure, you can follow the exact repro steps with the attached kernelbase.dll in the zip[
[KernelBase.zip](https://github.com/Vector35/binaryninja-api/files/12886356/KernelBase.zip)
](url)

**Comment by oldsouthcyber** [comment]
I am still having this issue. This is most notable when function parameters are not propogated throughout the function, and then there is another call later on and the subsequent function's param types are now lost). EG:

`myfunc1(typeX *arg1, typeY *arg2) {
....

NTSTATUS var1 = myfunc_sub(typeX* arg1,  typeY* arg2) {
var_2 = typeX* arg1;
var_3 = void* arg2; //2nd param has lost type
}

}`

**Comment by psifertex** [comment]
Can you share a specific example (including the binary) of it failing in a new issue? Ironically, I'm going to close the original issue since that particular instance is now resolved. 

When I attempt to reproduce it it looks correct (though I don't think that particular type is right, it does appear to be propagated):

```struct OBJECT_ATTRIBUTES** arg1```

```c
  int64_t PsspValidateWin32WalkMarker(struct OBJECT_ATTRIBUTES** arg1)

      struct OBJECT_ATTRIBUTES** arg_8 = arg1
      void var_128
      int64_t rax_1 = __security_cookie ^ &var_128
      int64_t rdx
      int64_t var_d8 = rdx
      
      if (arg1 != 0)
          int32_t rbx_1 = *arg1
          int32_t var_fc_1 = rbx_1
          
          if (rbx_1 != 0x4d575350)
              int32_t s
              memset(&s, c: 0, n: 0x98)
              s = 0xc0000008
              int32_t var_c4_1 = 0
              int64_t var_c0_1 = 0
              int64_t var_b8_1 = rdx
              int32_t var_b0_1 = 4
              struct OBJECT_ATTRIBUTES** var_a8_1 = arg1
              int64_t var_a0_1 = 0
              uint64_t var_98_1 = rbx_1
              int64_t var_90
              __builtin_strncpy(dest: &var_90, src: "PSWM", count: 8)
              RtlRaiseException(&s)
      
      return __security_check_cookie(rax_1 ^ &var_128)
```

**Comment by oldsouthcyber** [comment]
Yes, I am working on WdFilter.sys. Example would be DriverEntry:

DriverObject (arg1) is preserved in R14, then before functions such as _WPP_SF_i_ and _MpInitializeGlobals_ is is restored to RCX to be used as arg1 as, but the type is lost in the proceeding function definition. I can try to provide more concrete examples if required.

If possible I'd like to re-label this as a medium or high priority as it affects the decompilation significantly.

[WdFilter.zip](https://github.com/user-attachments/files/26121081/WdFilter.zip)
